DSA-Campaign-Uplift-Estimation/src/main/hangout_bot/messages.py

```diff
@@ -0,0 +1,487 @@
+            domains = message.split(',')
+            for idx in range(len(domains)):
+                # trim white space
+                domains[idx] = domains[idx].strip()
+                if not re.match(url_regex, domains[idx]):
+                    return failed
+            return success
+        
+        return success if re.match(url_regex, message) else failed
+    elif phase_num == 9:
+        if (len(message) > 0 and message != error_dictionary[0]):
+            return success
+        else:
+          return 'Ad text is not valid! Cannot be an empty value!'
+
+
+# [Message Generating Functions]
+
+
+def error_message(error_msg, phase_num):
+    """Formats an error message response for invalid input
+    Args:
+      error_msg:
+        the specifc error message being displayed
+      phase_num:
+        the phase number that raised the error
+    Returns:
+      dict
+        dictionary contains error response message
+    """
+
+    error = "<font color=\"#ff0000\">ERROR</font>: {}<br><b>Please send a valid value for your campaign {}!</b>".format(error_msg, phase_dictionary.get(phase_num)[PHASE_NAME_INDEX]) if phase_num != -1 else "<font color=\"#ff0000\">ERROR</font>: <b>{}</b>".format(error_msg)
+    return {
+              "cards": [
+                {
+                  "header": {
+                    "title": "Dynamic Search Ads Configuration Bot",
+                    "subtitle": "Alert",
+                    "imageStyle": "AVATAR",
+                    "imageUrl": "https://9to5google.com/2018/05/05/exclusive-new-google-app-icons-i-o-2018-gallery/ads_512dp/"
+                  },
+                  "sections": [
+                    {
+                      "widgets": [
+                        {
+                          "textParagraph": {
+                            "text": error
+                          }
+                        }
+                      ]
+                    }
+                  ]
+                }
+              ]
+            }
+
+
+def create_confirmation_message(event, phase_num, editing):
+    """Formats a confirmation message response for valid input
+    Args:
+      event:
+        the specifc event that has passed error
+        checking and is ready to be user confirmed
+      phase_num:
+        the phase number that will be used
+        to confirm the input value
+    Returns:
+      dict
+        dictionary contains confirmation prompt
+    """
+
+    return {
+              "cards": [
+                {
+                  "header": {
+                    "title": "Dynamic Search Ads Configuration Bot",
+                    "subtitle": "Editing",
+                    "imageStyle": "AVATAR",
+                    "imageUrl": "https://9to5google.com/2018/05/05/exclusive-new-google-app-icons-i-o-2018-gallery/ads_512dp/"
+                  },
+                  "sections": [
+                    {
+                      "widgets": [
+                        {
```

Comment:
Can you use a function to generate the boilerplate that is repeated across every message? That would make it simpler to read/edit in the future.

Reply:
For buttons, it will be a bit harder because some of the buttons require parameters and some dont. For now, I just added a function for the repeating header.